File: internal/runner/prompt.go
package runner

import (
	"fmt"

	"github.com/anubhavitis/pr-buddy/internal/artifact"
)

// Prompt builds the review instruction for a pull request.
//
// The rubric is versioned by RubricVersion; changing this text without bumping
// that constant would silently serve reviews produced under a different rubric.
func Prompt(p artifact.Provenance) string {
	return fmt.Sprintf(`Review pull request #%d in %s.

The working directory is a detached checkout of the pull request head (%s).
Its base is %s. Review only what this pull request changes: run
`+"`git diff %s...%s`"+` conceptually by reading the changed files, and judge the
change, not the pre-existing codebase.

You have read-only access. Do not attempt to modify files or run commands.

Report a finding only when you can point to the specific code that is wrong and
say what goes wrong because of it. A finding that would apply to almost any
codebase is not worth reporting. Prefer a short list of substantiated problems
over broad coverage.

Severity:
- error: a defect that will cause incorrect behaviour, data loss, or a security
  problem.
- warning: a real problem that is likely to cause trouble but is conditional or
  recoverable.
- info: worth the author's attention but not a defect.

Also produce a reading guide: an ordered set of groups that tells a reviewer
which files to read, in what order, and why. Order by what a reviewer must
understand first, not by file path.

Respond with a single JSON object and nothing else:

{
  "summary": "two or three sentences on what this change does and where the risk is",
  "findings": [
    {
      "severity": "error|warning|info",
      "rule": "short-kebab-case-category",
      "message": "what is wrong, in one sentence",
      "path": "path/relative/to/repo/root.go",
      "line": 42,
      "end_line": 45,
      "evidence": "the concrete reason: the input that breaks it, the conflicting call site, the unhandled branch",
      "confidence": 0.0,
      "reading_group": "name of the group this belongs to"
    }
  ],
  "reading_guide": [
    {
      "name": "group name",
      "summary": "why this group matters and what to look for",
      "paths": ["path/one.go", "path/two.go"]
    }
  ]
}

Use paths relative to the repository root. Set confidence between 0 and 1,
reflecting how sure you are the finding is real. Return an empty findings array
if the change is sound.`,
		p.PRNumber, p.Repo, shortSHA(p.HeadSHA), shortSHA(p.BaseSHA),
		shortSHA(p.BaseSHA), shortSHA(p.HeadSHA))
}

func shortSHA(s string) string {
	if len(s) > 12 {
		return s[:12]
	}
	return s
}

File: browser/src/prompt.ts
export type PromptFile = {
  path: string;
  additions?: number;
  deletions?: number;
};

export type PromptPR = {
  owner: string;
  repo: string;
  number: number;
  title: string;
  body: string;
  headSHA: string;
  files: PromptFile[];
};

export const PROMPT_TEMPLATE = `You are writing a reading order for a human reviewing a GitHub pull request.

Do not review the change. Do not list bugs. Order the files so the reviewer
understands the shape of the change before they read the wiring.

How to group
- 3 to 6 groups. Short names, two to four words (for example: Contracts, Then the host, Wiring).
- Read order: contracts / types / public API first, then implementation, then wiring, then tests.
- A group is files that should be read together, not a folder name.
- Put danger (auth, data, money, concurrency) in the file blurb, not in the sort key.

What to write
- Group summary: one line on why this cluster comes here.
- File blurb: two or three lines on what this file is doing in this PR.

This pull request
PR: {owner}/{repo}#{number}
Title: {title}
Description:
{description}

Changed files:
{files}

Reply with one JSON object and nothing else:

{
  "groups": [
    {
      "name": "group name",
      "summary": "why this group, what to look for",
      "files": [
        { "path": "path/relative/to/repo.go", "blurb": "two or three lines on what this file is" }
      ]
    }
  ]
}

Use only the paths listed above. Do not invent files. Do not report defects.`;

export function promptTemplate(): string {
  return PROMPT_TEMPLATE;
}

export function buildPrompt(pr: PromptPR): string {
  const files = pr.files
    .map((f) => {
      const plus = f.additions != null ? `+${f.additions}` : "";
      const minus = f.deletions != null ? `-${f.deletions}` : "";
      const stats = [plus, minus].filter(Boolean).join(" ");
      return stats ? `- ${f.path} ${stats}` : `- ${f.path}`;
    })
    .join("\n");

  const body = pr.body.trim() || "(no description)";

  return PROMPT_TEMPLATE.replaceAll("{owner}", pr.owner)
    .replaceAll("{repo}", pr.repo)
    .replaceAll("{number}", String(pr.number))
    .replaceAll("{title}", pr.title)
    .replaceAll("{description}", body)
    .replaceAll("{files}", files);
}
